File: lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
	public TreeNode LcaDeepestLeaves(TreeNode root) {
		HashSet<TreeNode> list = new HashSet<TreeNode>();
		Queue<TreeNode> queue = new Queue<TreeNode>();
		Dictionary<TreeNode, TreeNode> dict = new Dictionary<TreeNode, TreeNode>();

		queue.Enqueue(root);
		dict[root] = root;

		while(queue.Count > 0) {
			int size = queue.Count;
			list = new HashSet<TreeNode>();

			for(int i = 0; i < size; i++) {
				TreeNode node = queue.Dequeue();
				list.Add(node);

				if(node.left != null) {
					dict[node.left] = node;
					queue.Enqueue(node.left);
				}

				if(node.right != null) {
					dict[node.right] = node;
					queue.Enqueue(node.right);
				}
			}
		}

		while(list.Count > 1) {
			HashSet<TreeNode> parent = new HashSet<TreeNode>();
			foreach(var item in list)
				parent.Add(dict[item]);

			list = parent;
		}

		return list.First();
	}
}

File: 1218-lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def maxdepth(node,d): # to find depth of the tree T(n) S(h)
            d1,d2=0,0
            if not node:
                return d
            if node and node.left:
                d2=maxdepth(node.left,d+1)
            if node and node.right:
                d1=maxdepth(node.right,d+1)
            return max(d1,d2,d)
        self.depth=maxdepth(root,0) #S(1)
        self.lon=[] # S(h)
        def dfs(node,d,path): # T(n) S(h)
            if node and d==self.depth:
                path+=[node]
                if self.lon==[]:
                    self.lon=path
                    return
                i=0
                while i<len(self.lon): # runs for h time but after each iteration it becomes h-1 which can run for worst case of n/2 times
                    # this follows the pattern of sum of first n natural number but reverse so its h**2 complexity in worst case where its a complete binary tree with all levels filled
                    if self.lon[i]==path[i]:
                        i+=1
                    else:
                        self.lon=self.lon[:i]
                        break
            if not node:
                return
            if node and node.left:
                dfs(node.left,d+1,path+[node])
            if node and node.right:
                dfs(node.right,d+1,path+[node])
        dfs(root,0,[])
        return self.lon[-1]
